src/nspawn/nspawn.c

```diff
@@ -2226,97 +2227,120 @@ static bool should_enable_fuse(void) {
         return true;
 }
 
-static int copy_devnodes(const char *dest, bool enable_fuse) {
-        _cleanup_strv_free_ char **devnodes = NULL;
-        int r = 0;
+static int copy_devnode_one(const char *dest, const char *node, bool ignore_mknod_failure) {
+        int r;
 
         assert(dest);
+        assert(!isempty(node));
 
-        devnodes = strv_new("null",
-                            "zero",
-                            "full",
-                            "random",
-                            "urandom",
-                            "tty",
-                            STRV_IFNOTNULL(enable_fuse ? "fuse" : NULL),
-                            "net/tun");
-        if (!devnodes)
+        BLOCK_WITH_UMASK(0000);
+
+        _cleanup_free_ char *from = path_join("/dev/", node);
+        if (!from)
                 return log_oom();
 
-        BLOCK_WITH_UMASK(0000);
+        _cleanup_free_ char *to = path_join(dest, from);
+        if (!to)
+                return log_oom();
 
-        /* Create /dev/net, so that we can create /dev/net/tun in it */
-        if (userns_mkdir(dest, "/dev/net", 0755, 0, 0) < 0)
-                return log_error_errno(r, "Failed to create /dev/net directory: %m");
+        struct stat st;
+        if (stat(from, &st) < 0) {
+                if (errno != ENOENT)
+                        return log_error_errno(errno, "Failed to stat %s: %m", from);
 
-        STRV_FOREACH(d, devnodes) {
-                _cleanup_free_ char *from = NULL, *to = NULL;
-                struct stat st;
+                log_debug_errno(errno, "Device node %s does not exist, cannot copy it, ignoring.", from);
+                return 0;
```

Comment:
Is this the right thing to do?

The from device node may be missing in `/dev` due to `DevicePolicy=closed` and if that happens, we now give up immediately whereas before this MR, we'd attempt `mknod`. I would have expected the code to continue in the `ENOENT` case, but maybe I am getting things wrong here.

Reply:
I think this is a bug in the previous code. Previously in that case we passed uninitialized mode and rdev to mknod().

Reply:
The behavior is unchanged. We have called mknod() only when stat() succeeds.

Reply:
What you write makes sense, but it also makes me wonder whether my/our understanding of my regression report is correct. The observation was (I checked production logs again, ellipsis is mine):

    mknod(/.../dev/net/tun) failed: Operation not permitted

My understanding is that `DevicePolicy=closed` would cause the `stat` on `/dev/net/tun` to yield `ENOENT`. You reasonably argue that we never called `mknod` in that case. The production log appears to contradict that reasoning, so one of the premises is wrong and I am unsure which. Do you see which?

Reply:
Even when DevicePolicy=closed without DeviceAllow=, calling stat() for /dev/net/tun succeeds, but mknod() will fail, and bind mount will succeed. You can see by the following:
- without --private-users
```
$ sudo systemd-run --pty --wait -p Environment=SYSTEMD_LOG_LEVEL=debug -p DevicePolicy=closed -p DeviceAllow="char-pts rw" ./systemd-nspawn -xD / --private-network
(snip)
Failed to open /dev/fuse: Operation not permitted
Disabling FUSE: Failed to determine FUSE version: Operation not permitted
Bind-mounting /dev/net/tun on /.#machine.991d52b7ef6a1129/dev/net/tun (MS_BIND "")...
Bind-mounting /run/systemd/nspawn/unix-export/thinkpad-x280-34bf04db33dd233b on /.#machine.991d52b7ef6a1129/run/host/unix-export (MS_BIND "")...
(snip)
```
- with --private-users=pick
```
$ sudo systemd-run --pty --wait -p Environment=SYSTEMD_LOG_LEVEL=debug -p DevicePolicy=closed -p DeviceAllow="char-pts rw" ./systemd-nspawn -xD / --private-network --private-users=pick
(snip)
Failed to open /dev/fuse: Operation not permitted
Disabling FUSE: Failed to determine FUSE version: Operation not permitted
Failed to mknod(/.#machine.88f9129719dc2877/dev/net/tun): Operation not permitted
(sd-namespace) failed with exit status 1.
Finished with result: exit-code
Main processes terminated with: code=exited/status=1
Service runtime: 387ms
CPU time consumed: 171ms
Memory peak: 2.4M
Memory swap peak: 0B
```

Reply:
Also see the added test cases.